File: app/src/main/java/ipvc/estg/incidentes/services/MarkerClusterRenderer.java
package ipvc.estg.incidentes.services;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.core.content.ContextCompat;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.model.BitmapDescriptorFactory;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.MarkerOptions;
import com.google.maps.android.clustering.ClusterManager;
import com.google.maps.android.clustering.view.DefaultClusterRenderer;
import com.google.maps.android.ui.IconGenerator;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import ipvc.estg.incidentes.R;
import ipvc.estg.incidentes.constructors.MyMarker;

@SuppressLint("InflateParams")
public class MarkerClusterRenderer extends DefaultClusterRenderer<MyMarker> implements GoogleMap.OnInfoWindowClickListener {

    private GoogleMap googleMap;
    private LayoutInflater layoutInflater;
    private static final int MARKER_DIMENSION = 48;
    private final IconGenerator iconGenerator;
    private final ImageView markerImageView;
    private Context mycontext;


    public MarkerClusterRenderer(@NonNull Context context, GoogleMap map, ClusterManager<MyMarker> clusterManager) {
        super(context, map, clusterManager);
        mycontext = context;
        iconGenerator = new IconGenerator(context);
        markerImageView = new ImageView(context);
        markerImageView.setLayoutParams(new ViewGroup.LayoutParams(MARKER_DIMENSION, MARKER_DIMENSION));
        iconGenerator.setContentView(markerImageView);
        this.googleMap = map;
        layoutInflater = LayoutInflater.from(context);
        googleMap.setInfoWindowAdapter(clusterManager.getMarkerManager());
        googleMap.setOnInfoWindowClickListener(this);
        clusterManager.getMarkerCollection().setOnInfoWindowAdapter(new MyCustomClusterItemInfoView());
        googleMap.setOnCameraIdleListener(clusterManager);
        googleMap.setOnMarkerClickListener(clusterManager);
    }

    @Override
    protected void onBeforeClusterItemRendered(MyMarker item, MarkerOptions markerOptions) {
        if(item.getStatus() == 1){
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_launcher_background));
        }else if(item.getStatus() == 2 || item.getStatus() == 3){
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_launcher_background));
        }else if(item.getStatus() == 0){
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_launcher_background));
        }else{
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_launcher_background));
        }
    }

    @Override
    protected void onClusterItemRendered(MyMarker clusterItem, Marker marker) {
        marker.setTag(clusterItem);
    }

    @Override
    public void onInfoWindowClick(Marker marker) {
        MyMarker myMarker = (MyMarker) marker.getTag();
       /* Intent intent = new Intent(mycontext, SingleActivity.class);
        intent.putExtra("position",myMarker.getPosition());
        intent.putExtra("latitude",myMarker.getPosition().latitude);
        intent.putExtra("longitude",myMarker.getPosition().longitude);
        intent.putExtra("location",myMarker.getLocation());
        intent.putExtra("date",myMarker.getDate());
        intent.putExtra("time",myMarker.getTime());
        intent.putExtra("status",myMarker.getStatus());
        intent.putExtra("description",myMarker.getDescription());
        intent.putExtra("id",myMarker.getId());
        intent.putExtra("number",myMarker.getNumber());
        intent.putExtra("photo",myMarker.getPhoto());
        intent.putExtra("photo_finish",myMarker.getPhoto_finish());

        mycontext.startActivity(intent);*/

    }

    private class MyCustomClusterItemInfoView implements GoogleMap.InfoWindowAdapter {

        private final View clusterItemView;

        MyCustomClusterItemInfoView() {
            clusterItemView = layoutInflater.inflate(R.layout.custominfowindow, null);
        }

        @Override
        public View getInfoWindow(final Marker marker) {
            final MyMarker myMarker = (MyMarker) marker.getTag();
            if (myMarker == null) return clusterItemView;

            TextView status = clusterItemView.findViewById(R.id.status);
            Log.d("STATUS", String.valueOf(myMarker.getStatus()));
            if(myMarker.getStatus() == 1){
                clusterItemView.findViewById(R.id.number_color).setBackgroundColor(ContextCompat.getColor(mycontext, R.color.cpb_green));
                status.setText("completo");
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.cpb_blue));
            }else if(myMarker.getStatus() == 2 || myMarker.getStatus() == 3){
                clusterItemView.findViewById(R.id.number_color).setBackgroundColor(ContextCompat.getColor(mycontext, R.color.cpb_red));
                status.setText("progress");
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.cpb_red));
            }else if(myMarker.getStatus() == 0){
                clusterItemView.findViewById(R.id.number_color).setBackgroundColor(ContextCompat.getColor(mycontext, R.color.cpb_green_dark));
                status.setText("recived");
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.cpb_red_dark));
            }else{
                clusterItemView.findViewById(R.id.number_color).setBackgroundColor(ContextCompat.getColor(mycontext, R.color.cpb_grey));
                status.setText("error");
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.cpb_grey));
            }




            TextView name = clusterItemView.findViewById(R.id.name);
            TextView date = clusterItemView.findViewById(R.id.date);
            TextView location = clusterItemView.findViewById(R.id.location);
       /*     TextView description = clusterItemView.findViewById(R.id.description);*/
            name.setText(myMarker.getNumber());

            location.setText(myMarker.getLocation());
        /*    description.setText(myMarker.getDescription());*/

            try {
                Date date_time = new SimpleDateFormat("dd/MM/yyyy").parse(myMarker.getDate());
                String formattedDate = new SimpleDateFormat("dd MMMM yyyy").format(date_time);
                date.setText(formattedDate+ " "+myMarker.getTime());
            } catch (ParseException e) {
                e.printStackTrace();
            }
            return clusterItemView;
        }

        @Override
        public View getInfoContents(final Marker marker) {

            return null;
        }
    }
}

File: app/src/main/java/ipvc/estg/incidentes/services/MarkerClusterRenderer.kt
package ipvc.estg.incidentes.services

import android.annotation.SuppressLint
import android.content.Context
import android.content.SharedPreferences
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.InfoWindowAdapter
import com.google.android.gms.maps.GoogleMap.OnInfoWindowClickListener
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.maps.android.clustering.Cluster
import com.google.maps.android.clustering.ClusterManager
import com.google.maps.android.clustering.view.DefaultClusterRenderer
import com.google.maps.android.ui.IconGenerator
import ipvc.estg.incidentes.R
import ipvc.estg.incidentes.entities.MyMarker
import ipvc.estg.incidentes.fragments.EventDetailFragment
import ipvc.estg.incidentes.fragments.HomeFragment
import ipvc.estg.incidentes.navigation.NavigationHost
import java.text.ParseException
import java.text.SimpleDateFormat


@SuppressLint("InflateParams")
class MarkerClusterRenderer(
    private val mycontext: Context,
    map: GoogleMap,
    clusterManager: ClusterManager<MyMarker?>
) : DefaultClusterRenderer<MyMarker>(mycontext, map, clusterManager), OnInfoWindowClickListener {
    private val googleMap: GoogleMap
    private val layoutInflater: LayoutInflater
    private val iconGenerator: IconGenerator = IconGenerator(mycontext)
    private val markerImageView: ImageView = ImageView(mycontext)
    private val logged: Boolean = (mycontext as NavigationHost).isUserLogged() == false
    private val idUser: Int = (mycontext as NavigationHost).getAuthenticationUserId()!!
    override fun onBeforeClusterItemRendered(item: MyMarker, markerOptions: MarkerOptions) {
        if (item.status.id == 1) {
            if(logged && idUser == item.user_id){

            }else{
                markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_green_pin))
            }
        } else if (item.status.id == 2 || item.status.id == 3) {
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_yellow_pin))
        } else if (item.status.id == 4) {
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_red_pin))
        } else {
            markerOptions.icon(BitmapDescriptorFactory.fromResource(R.drawable.ic_green_pin))
        }
    }

    override fun onClusterItemRendered(clusterItem: MyMarker, marker: Marker) {
        marker.tag = clusterItem
    }

    override fun onInfoWindowClick(marker: Marker) {
        val myMarker = marker.tag as MyMarker?
        val bundle = Bundle()
        bundle.putString("destination", "view")
        bundle.putInt("id", myMarker!!.id)
        bundle.putCharSequence("position", myMarker.position.toString())
        bundle.putDouble("latitude", myMarker.position.latitude)
        bundle.putDouble("longitude", myMarker.position.longitude)
        bundle.putString("date", myMarker.date)
        bundle.putString("time", myMarker.time)
        bundle.putInt("status", myMarker.status.id)
        bundle.putString("description", myMarker.description)
        bundle.putString("number", myMarker.number)
        bundle.putString("photo", myMarker.photo)
        bundle.putString("photo_finish", myMarker.photo_finish)
        bundle.putBoolean("owner", myMarker.user_id == idUser)
        bundle.putInt("idUser", idUser)
        bundle.putString("action", "view")
        bundle.putInt("type", myMarker.type)

        (mycontext as NavigationHost).navigateToWithData(
            EventDetailFragment(),
            addToBackstack = true,
            animate = true,
            data = bundle,
            tag = "event_detail"
        )
    }


    override fun shouldRenderAsCluster(cluster: Cluster<MyMarker>?): Boolean {
        val preferences: SharedPreferences = mycontext.getSharedPreferences("MAPZOOM", Context.MODE_PRIVATE)
        val zoom = preferences.getFloat("zoom", 0.0f)
        if(zoom >= 19.5f && zoom != 0.0f){
            return false
        }
        return cluster!!.size > 1
    }


    private inner class MyCustomClusterItemInfoView internal constructor() : InfoWindowAdapter {
        private val clusterItemView: View

        override fun getInfoWindow(marker: Marker): View {
            val myMarker = marker.tag as MyMarker? ?: return clusterItemView
            val status = clusterItemView.findViewById<TextView>(R.id.status)
            if (myMarker.status.id == 1) {
                clusterItemView.findViewById<View>(R.id.number_color).setBackgroundColor(
                    ContextCompat.getColor(
                        mycontext,
                        R.color.success
                    )
                )
                status.setText(R.string.status_completed)
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.success))
            } else if (myMarker.status.id == 2 || myMarker.status.id == 3) {
                clusterItemView.findViewById<View>(R.id.number_color).setBackgroundColor(
                    ContextCompat.getColor(
                        mycontext,
                        R.color.warning
                    )
                )
                status.setText(R.string.status_in_progress)
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.warning))
            } else if (myMarker.status.id == 4) {
                clusterItemView.findViewById<View>(R.id.number_color).setBackgroundColor(
                    ContextCompat.getColor(
                        mycontext,
                        R.color.danger
                    )
                )
                status.setText(R.string.status_received)
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.danger))
            } else {
                clusterItemView.findViewById<View>(R.id.number_color).setBackgroundColor(
                    ContextCompat.getColor(
                        mycontext,
                        R.color.cpb_grey
                    )
                )
                status.setText(R.string.status_error)
                status.setTextColor(ContextCompat.getColor(mycontext, R.color.cpb_grey))
            }
            val name = clusterItemView.findViewById<TextView>(R.id.name)
            val date = clusterItemView.findViewById<TextView>(R.id.date)
            val location = clusterItemView.findViewById<TextView>(R.id.location)
            val myReport = clusterItemView.findViewById<LinearLayout>(R.id.my_report)
            name.text = myMarker.location
            location.text = myMarker.location

            if(myMarker.user_id == idUser){
                myReport.visibility = View.VISIBLE
            }else{
                myReport.visibility = View.GONE
            }

            try {
                val creationDate = SimpleDateFormat("dd/MM/yyyy").parse(myMarker.date)
                val day = SimpleDateFormat("dd MMMM").format(creationDate)
                val year = SimpleDateFormat("yyyy").format(creationDate)
                date.text = mycontext.resources.getString(R.string.date, day, year, myMarker.time)
            } catch (e: ParseException) {
                e.printStackTrace()
            }
            return clusterItemView
        }

        override fun getInfoContents(marker: Marker): View? {
            return null
        }

        init {
            clusterItemView = layoutInflater.inflate(R.layout.custominfowindow, null)
            clusterItemView
        }
    }

    companion object {
        private const val MARKER_DIMENSION = 48
    }

    init {
        markerImageView.layoutParams = ViewGroup.LayoutParams(MARKER_DIMENSION, MARKER_DIMENSION)
        iconGenerator.setContentView(markerImageView)
        googleMap = map
        layoutInflater = LayoutInflater.from(mycontext)
        googleMap.setInfoWindowAdapter(clusterManager.markerManager)
        googleMap.setOnInfoWindowClickListener(this)
        clusterManager.markerCollection.setOnInfoWindowAdapter(MyCustomClusterItemInfoView())
        googleMap.setOnCameraIdleListener(clusterManager)
        googleMap.setOnMarkerClickListener(clusterManager)
    }
}